extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchBufferAggregator.java

```diff
@@ -87,12 +81,27 @@ public Object get(ByteBuffer buf, int position)
   //Note that this is not threadsafe and I don't think it needs to be
   private Union getUnion(ByteBuffer buf, int position)
   {
-    Union union = unions.get(position);
-    if (union == null) {
-      Memory mem = new MemoryRegion(nm, position, maxIntermediateSize);
-      union = (Union) SetOperation.wrap(mem);
-      unions.put(position, union);
+    Int2ObjectMap<Union> unionMap = unions.get(buf);
+    Union union = unionMap != null ? unionMap.get(position) : null;
+    if (union != null) {
+      return union;
+    }
+    return createNewUnion(buf, position, true);
+  }
+
+  private Union createNewUnion(ByteBuffer buf, int position, boolean isWrapped)
+  {
+    NativeMemory nm = getNativeMemory(buf);
+    Memory mem = new MemoryRegion(nm, position, maxIntermediateSize);
+    Union union = isWrapped
+                  ? (Union) SetOperation.wrap(mem)
+                  : (Union) SetOperation.builder().initMemory(mem).build(size, Family.UNION);
+    Int2ObjectMap<Union> unionMap = unions.get(buf);
```

Comment:
I guess your idea is that the contract would be something like this:

- `init` is called before starting aggregation. all buffer objects passed in should be the same buffer.
- `aggregate` is called to aggregate. all buffer objects passed in should be the same buffer.
- `relocate` is called to move to a new buffer, or new location within a buffer. It's expected that for all calls to relocate: (1) the `oldBuffer`s are the same from call to call, and the `newBuffer`s are also the same from call to call; (2) there are no intervening calls to `aggregate`.
- after relocate is done, `aggregate` can continue being called, but the buffer must now be the new buffer (i.e. the buffer that was passed as `newBuffer` to all relocate calls).

I think with that, BufferGrouper can work, and also SketchBufferAggregator wouldn't need to keep maps.

I'm still willing to defer to @akashdw whether or not this optimization is necessary. If the `SketchAggregator.updateUnion` is expected to take seriously longer than `getUnion` then the optimization might not yield a noticeable performance difference.